### ISSUE TITLE
non-interactive composer2 config --auth

### ISSUE THREAD

**Issue by drzraf** [composer/composer#9731]
Output of `composer diagnose`:

```
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Composer version: 2.0.9
PHP version: 7.4.14
PHP binary path: /usr/bin/php7.4
OpenSSL version: OpenSSL 1.1.1i  8 Dec 2020
cURL version: 7.64.0 libz 1.2.11 ssl OpenSSL/1.1.1i
```

When I run:
```
composer config -f ~/.config/composer/auth.json  gitlab-token.gitlab.com xxx
```

I get the following output:

```
  [RuntimeException]                                                                    
  File "/home/xxx/.config/composer/auth.json" cannot be found in the current directory 
```

And I expected this to happen:
- `~/.config/composer` to be created
- `~/.config/composer/auth.json` to be created with the necessary credentials

Actually I would even expect the following to work:
`composer config -ga gitlab-token.gitlab.com xxx`

But it's even worst. `composer2` says directory must pre-exist, lets `mkdir ~/.config/composer` and run the command again.
```
  [RuntimeException]                                                                    
  File "/home/xxx/.config/composer/auth.json" cannot be found in the current directory 
```

Ok, lets `touch ~/.config/composer/auth.json` and run the command again.
```
  [Seld\JsonLint\ParsingException]                                        
  "/home/xxx/.config/composer/auth.json" does not contain valid JSON      
  Parse error on line 1:
```

Basically composer2 is not willing to setup the authentication using the usual `-g / -a` (problem 1) and force a file path to be specified.
If said file does not exist it's not created (problem 1 bis)
If said file is created by other mean (up to the user then), it's likely to not work at all (problem 2), totally breaking composer usage for that specific user (problem 3).

Authentication token are likely involved in organization management and could easily part of infrastructure/machine provisioning. Setting up authentication non-interactively is a must-have.

**Comment by Seldaek** [comment]
I can't repro this..

```
$ rm -rf ~/.config/composer/
$ composer config -ga gitlab-token.gitlab.com xxx
$ cat ~/.config/composer/auth.json
{
    "bitbucket-oauth": {},
    "github-oauth": {},
    "gitlab-oauth": {},
    "gitlab-token": {
        "gitlab.com": "xxx"},
    "http-basic": {},
    "bearer": {}
}
```

So not sure what you are doing but it seems like there's something else in your environment which is causing this.

**Comment by drzraf** [comment]
```
$ rm -rf ~/.config/composer/
$ composer config -ga gitlab-token.gitlab.com xxx
                                              
  [RuntimeException]                                                                                                            
  Failed to update composer.json with a valid format, reverting to the original content. Please report an issue to us with det  
  ails (command you run and a copy of your composer.json).                               
                                                                             
  [Composer\Json\JsonValidationException]                                    
  "/home/xxx/.composer/config.json" does not match the expected JSON schema
```

But **I do have an existing** `~/.composer` directory (including `config.json` and `auth.json`)

**Comment by stof** [comment]
> `config.json` and `auth.json` files are actually created but ... in `~/.composer` instead of `~/.config/composer`.

This depends whether your system uses the XDG Base Directory Specifications or no.

**Comment by drzraf** [comment]
Two Debian 10 (Buster), both provisioned by the same Ansible playbooks (with the same user environment variables):
`rm -f .composer/*.json .config/composer/*json; composer config -ga gitlab-token.gitlab.com xxx && ls -l .composer/*json .config/composer`

- `machine 1`: 
`.composer/auth.json  .composer/config.json`

- `machine 2`: 
`.config/composer/auth.json .config/composer/config.json`

**Comment by stof** [comment]
Well, on systems that follow XDG, composer has 2 choices for its config: `$XDG_CONFIG_HOME/composer` (i.e. `~/.config/composer` if using the XDG default config home) and `~/.composer (which is the older convention supported by composer). To avoid disruption when XDG support was added, composer will use the first directory that exists (and use the first one, i.e. the XDG one if none already exist).
If you want to go back to a clean state to restart the configuration of composer, remove both folders (to account for a system where the legacy `~/.composer` was used).

**Comment by Seldaek** [comment]
Closing as I don't really see how we can improve this without causing issues for other cases.

**Comment by drzraf** [comment]
My suggestion: Improve documentation so that when one reads the above quoted sentence about `whether your system uses the XDG Base Directory Specifications or no.`, you could add something like:

> ...  according to the existence of **TWO** directories which are checked for: `$XDG_CONFIG_HOME/composer` **AND** `$XDG_CACHE_HOME/composer`, usually: `~/.config/composer/` and `~/.config/composer/`)
